src/compiler/propagation/type_database.cc

```diff
@@ -36,10 +41,97 @@ TypeDatabase::~TypeDatabase() {
   }
 }
 
+void TypeDatabase::check_top(uint8* bcp, Object* value) const {
+  int position = program_->absolute_bci_from_bcp(bcp);
+  auto probe = usage_.find(position);
+  if (probe == usage_.end()) {
+    FATAL("usage not analyzed: %d", position);
+  }
+  TypeSet type = probe->second;
+  if (type.is_block()) {
+    // TODO(kasper): We should improve the type check
+    // for blocks and verify that they point into the
+    // right stack section.
+    if (is_smi(value)) return;
+    FATAL("expected a block at %d", position);
+  }
+  Smi* class_id = is_smi(value)
+      ? program_->smi_class_id()
+      : HeapObject::cast(value)->class_id();
+  if (type.contains(class_id->value())) return;
+  FATAL("didn't expect %d at %d", class_id->value(), position);
+}
+
+void TypeDatabase::check_return(uint8* bcp, Object* value) const {
+  // TODO(kasper): This isn't super nice, but we have to avoid
+  // getting hung up over the intrinsic bytecodes. We sometimes
+  // return from a block and restart at the intrinsic bytecode,
+  // but we don't care about that for now. We could make the
+  // propagator allow any value as the top stack element here,
+  // but it would achieve the same things as this check.
+  uint8 opcode = *bcp;
+  if (opcode > HALT) return;
```

Comment:
Might be easier to read if you replace the `opcode > HALT` with:
`opcode >= INTRINSIC_SMI_REPEAT`.
You could have a static assert that `HALT + 1 == INTRINSIC_SMI_REPEAT` and that `INTRINSIC_HASH_DO + 1 == ILLEGAL_END`, and maybe even that there are only 4 of them.
That should make sure that we don't accidentally hit this `if` for the wrong bytecodes if we reshuffle them.